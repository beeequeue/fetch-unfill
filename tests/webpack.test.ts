/* eslint-disable unicorn/prefer-module */
import assert from "assert"
import path from "path"

import { createFsFromVolume, Volume } from "memfs"
import { describe, expect, test } from "vitest"
import { Configuration, webpack } from "webpack"
import { merge } from "webpack-merge"

const build = async (name: string, extraOptions: Configuration | null = {}) => {
  const compiler = webpack(
    merge(
      {
        entry: path.resolve(__dirname, "fixtures", `${name}.js`),
        resolve: {
          modules: [path.resolve(__dirname, "..", "node_modules")],
        },
        output: { libraryTarget: "commonjs-module" },

        mode: "production",
        target: "node18",
        stats: "none",
        devtool: false,
        experiments: { futureDefaults: true },
      },
      extraOptions ?? {},
    ),
  )

  const output = createFsFromVolume(new Volume())
  compiler.outputFileSystem = output

  return await new Promise<string | Error>((resolve, reject) =>
    compiler.run((err, stats) => {
      if (err || stats?.hasErrors()) {
        return reject(err ?? stats!.toJson({ errors: true }).errors?.[0])
      }

      resolve(output.readFileSync("dist/main.js", { encoding: "utf8" }) as string)
    }),
  )
}

const alias = {
  resolve: {
    alias: {
      "node-fetch": path.resolve(__dirname, "..", "lib", "index.cjs"),
      "cross-fetch": path.resolve(__dirname, "..", "lib", "index.cjs"),
    },
  },
} satisfies Configuration

const cases: Array<[name: string, snapshotName: string, options: Configuration | null]> =
  [
    ["bundles the original package", "webpack_node-fetch_bundled.js", null],
    ["unfills it", "webpack_node-fetch_unfilled.js", alias],
  ]

describe("node-fetch", () => {
  test.each(cases)(
    "%s",
    async (_, snapshotName, options) => {
      const result = await build("node-fetch", options)

      expect(result).not.toBeInstanceOf(Error)
      assert(!(result instanceof Error))

      await expect(result).toMatchFileSnapshot(`__snapshots__/${snapshotName}`)

      await expect(result).toBeAbleToFetch()
    },
    { timeout: 10_000 },
  )
})

describe("cross-fetch", () => {
  test.each(cases)(
    "%s",
    async (_, snapshotName, options) => {
      const result = await build("cross-fetch", options)

      expect(result).not.toBeInstanceOf(Error)
      assert(!(result instanceof Error))

      await expect(result).toMatchFileSnapshot(
        `__snapshots__/${snapshotName.replace("node-", "cross-")}`,
      )

      await expect(result).toBeAbleToFetch()
    },
    { timeout: 10_000 },
  )
})
