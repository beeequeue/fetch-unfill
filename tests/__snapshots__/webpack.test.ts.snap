// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cross-fetch > bundles the original package 1`] = `"86.84KB"`;

exports[`cross-fetch > unfills it 1`] = `"569B"`;

exports[`node-fetch > bundles the original package 1`] = `"85.36KB"`;

exports[`node-fetch > unfills it 1`] = `"569B"`;
